syntax = "proto3";

package seigr_protocol;

// Primary message representing a complete lineage for a Seigr object
message Lineage {
    string creator_id = 1;                     // ID of the creator initiating this lineage
    string current_hash = 2;                   // Current hash representing the latest state in the lineage
    string version = 3;                        // Version of the lineage format
    repeated LineageEntry entries = 4;         // Collection of lineage entries tracking each action
}

// A single entry in the lineage, representing a discrete action or state change
message LineageEntry {
    string version = 1;                        // Version of the entry format
    string action = 2;                         // Description of the action taken (e.g., "created", "modified")
    string creator_id = 3;                     // ID of the primary creator
    string contributor_id = 4;                 // ID of the contributor associated with this action
    string timestamp = 5;                      // Timestamp of the entry for chronological tracking

    // List of previous hashes, supporting multiple references for non-linear or branched lineage tracking
    repeated string previous_hashes = 6;

    // Flexible metadata map for storing additional context or key-value pairs related to the entry
    map<string, string> metadata = 7;
}
